src/repr/src/adt/timestamp.rs

```diff
@@ -704,6 +795,24 @@ impl<T: TimestampLike> CheckedTimestamp<T> {
         // If at any point we overflow, map to a TimestampError.
         age_inner(self, other).ok_or(TimestampError::OutOfRange)
     }
+
+    /// Rounds the timestamp to the specified number of digits of precision.
+    pub fn round_to_precision(&mut self, precision: Option<TimestampPrecision>) {
+        if let Some(precision) = precision {
+            let power = 9_u32
+                .checked_sub(precision.into_u8().into())
+                .expect("precision fits in nanos");
+            let round_to = Duration::from_std(StdDuration::from_nanos(10_u64.pow(power)))
+                .expect("duration fits in chrono duration");
+            let sec = Duration::from_std(StdDuration::from_secs(1))
+                .expect("duration fits in chrono duration");
+            let dt = self.date_time();
+            let dt = dt + sec;
+            let dt = dt.duration_round(round_to).expect("rounding is valid");
+            let dt = dt - sec;
+            self.t = T::from_date_time(dt)
```

Comment:
can you add a comment explaining this logic?

Reply:
Looks like this whole function will need to be rewritten to not use chrono's `duration_round`. That implementation is just inexplicably panicky in a whole bunch of situations where there's no need to panic. (The +1/-1 was a quick hack around one of the issues; unfortunately @nrainer-materialize found yet another panic above.) Shouldn't be too hard to write this logic manually by pulling out the subsec nanos and doing the modulus/addition/subtraction dance.

Reply:
I will update this PR to do the operations manually, will also add more scenarios to the testdrive tests.

Reply:
Updated this logic. I am using the same logic as the `duration_round` but in microsecond precision (since that's the highest) instead of nanoseconds which was frequently overflowing.

Also added unit tests with the lowest and highest dates.

Reply:
I removed the +/- of the second, since it was working with both the lowest and highest date value.

Reply:
@nrainer-materialize The statements above should work now.

Reply:
Excellent!

Reply:
Retested successfully!